content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,157 @@
+    
+![11_embeddings](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/11_embeddings.png)
+    
+
+Currently, [we default to a mix of both](https://github.com/sourcegraph/cody/blob/0238df0eda7e6e9be03139a56a89e235b76be448/vscode/package.json#L967) with an option to use one or the other.
+
+### Recent improvements to context fetching for chat
+
+With [the June release](https://sourcegraph.com/blog/release/june-2024), we made several improvements: Enterprise context now rewrites queries, enabling us to handle longer and foreign-language questions. We also detect filenames and symbols in queries, addressing a class of "easy-looking" questions users expect us to get right.
+
+---
+
+## Context sources outside of your codebase
+
+In addition to automatic context fetching, we've found that providing a way for the user to manually include context is also helpful. We've done this by letting the user ask questions about specific files by @-mentioning them. More recently, we've realized that there is more context the user sometimes wants to include outside of the codebase, such as documentation, web pages, Slack threads, Linear issues, and Notion documents. We've created an open source protocol called [OpenCtx](https://openctx.org/) to make it super easy to bring in context from any source.
+
+[![20_tweet](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/20_tweet.png)](https://x.com/beyang/status/1792616243742244974)
+
+
+---
+
+## Context fetching for autocomplete
+
+Autocomplete provides code suggestions as you code in your editor. For this feature, we want it to feel as instantaneous as possible because we don’t want to break the flow of the programmer. At the same time, we want to quickly find and identify which code snippets to include as part of the context to provide high-quality suggestions.
+
+![13_autocomplete](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/13_autocomplete.png)
+
+For this, we look at a few sources of information: the cursor position, the surrounding code, and the code graph (a code graph is a representation of the relationships and structures within a codebase, mapping entities such as classes and methods to show how they are interconnected). We use the current cursor position within the code graph to determine if the user wants a single-line suggestion or a multi-line suggestion. Once we determine that, we add more context by looking through recent files and open tabs. Within those, we find code snippets related to the code you're currently writing.
```

Comment:
It'd be good to explain how we use the code graph for autocomplete context. I know this is unclear to many people (both inside and out of Sourcegraph).

Reply:
Added a link to this to answer it - hope it clarifies it: https://sourcegraph.notion.site/How-do-we-use-the-code-graph-for-autocomplete-context-11145ad3e07049e8ab1add7ac3012f81

Reply:
That indeed helps. However It's still unclear to me when and if we actually use this. For example, is LSP context actually enabled by default for our customers?

Reply:
I looked into it, and it looks like it's determined by a feature flag: https://github.com/sourcegraph/cody/blob/main/vscode/src/completions/completion-provider-config.ts#L77

As well as a config variable: https://github.com/sourcegraph/cody/blob/54b25dc7d17bac425c2b59b04ca08d600ef184ed/vscode/package.json#L1137